File: performance/throughput/throughput_coniql.js
// node throughput_coniql.js
const { createClient } = require("graphqurl");
const base64js = require("base64-js");

const TEST_SUBSCRIPTION_URL = "ws://localhost:8080/ws";
const PV = "test:waveform";

let count = 0;
let skip = 0;
let t1 = process.hrtime();
let process_results = true;

const onConnection = () => {
    console.log('Connected to '+TEST_SUBSCRIPTION_URL)
    client.subscribe(
    {
      subscription: `subscription {subscribeChannel(id: "test:waveform") {id value{base64Array{base64}}}}`,
    },
    eventCallback,
    errorCallback
  )
}

const client = createClient({
    websocket: {
        endpoint: TEST_SUBSCRIPTION_URL,
        onConnectionSuccess: onConnection,
        onConnectionError: () => console.log('Connection Error'),
    }
});

const eventCallback = (event) => {
    if (count == 0 && skip < 2) {
        // Skip to first 2 which will be the initial connection
        // updates. Make sure we are just getting the value
        // updates
        skip = skip + 1
        console.log("skip first 2")
    } else {
        if (process_results) {
            const bd = base64js.toByteArray(event.data.subscribeChannel.value.base64Array.base64);
            const numbers = new Float64Array(bd.buffer);
            if (count == 99) {
                console.log(numbers.length)
            }
        }
        if (count == 0 ) {
            this.t1 = process.hrtime();
        }
        if (count == 99) {
            const t2 = process.hrtime(this.t1);
            const executionTime = (t2[0] * 1000 + t2[1] / 1000000) / 1000;
            console.log(`Final count: ${count}`);
            console.log(`Execution Time: ${executionTime}`);
            const messageFreq = count / executionTime;
            console.info(`Measured frequency: ${messageFreq} Hz`);
        }
        count = count + 1;
    }
};

const errorCallback = (error) => {
  console.log('Error:', error)
};

File: performance/throughput/throughput_coniql.py
import asyncio
import time
import websockets
from graphql_client import GraphQLClient
import argparse
import base64
import json
import numpy as np

parser = argparse.ArgumentParser()
parser.add_argument("-p", "--process",action="store_true")
args = parser.parse_args()


TEST_SUBSCRIPTION_URL = "ws://localhost:8080/ws"

sinewave_subscription = f"""subscription {{
  subscribeChannel(id: "test:waveform") {{
    id
    value {{
        base64Array {{
          base64
        }}
    }}
  }}
}}
"""

def to_float_array(input_data):
    buffer = base64.decodebytes(input_data.encode("ascii"))
    return np.frombuffer(buffer, dtype=np.float64)

def decode(response, check_length):
    if check_length:
        print("Array length: ", len(to_float_array(response["payload"]["data"]["subscribeChannel"]["value"]["base64Array"]["base64"])))



count = 0
def add_count(id, data):
    if args.process:
        decode(data, True)
    global count
    count += 1

def main():
    with GraphQLClient('ws://localhost:8080/ws') as ws:
        sub_id = ws.subscribe(query=sinewave_subscription, callback=add_count)

        start_time = time.time()

        time.sleep(30)

        ws.stop_subscribe(sub_id)

        end_time = time.time()
        ws.close()

        recorded_time = end_time - start_time
        print(f"Final count: {count}")
        print(f"Execution Time: {recorded_time} s")
        message_freq = count / recorded_time
        print(f"Measured Frequency: {message_freq} Hz")


if __name__ == "__main__":
    main()
    #asyncio.get_event_loop().run_until_complete(main())